### PR TITLE
feat: Add APM Server version fetching and client.supportsKeepingUnsampledTransaction()

### PR DESCRIPTION
**Pull request by trentm**
This is to support the Node.js APM agent changing its behaviour based on
the APM Server version. Specifically whether to send or drop unsampled
transactions.

Refs: https://github.com/elastic/apm-agent-nodejs/issues/2455

### CODE REVIEW

**Comment by trentm** [comment]
The first commit here is a first stab that has the APM Server fetching *retry* if it fails on earlier attempts. IIUC the java agent does *not* retry (see https://github.com/elastic/apm-agent-java/blob/master/apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerHealthChecker.java). I think I am going to remove the retrying in the node.js agent, but I am not sure yet. If there are no retries, then if the APM Server happens to be down when the agent starts, then the agent will stay in a "degraded" state not knowing the APM server version.  If this "degraded" state just assumes the latest APM Server version then that is *perhaps* fine, but it could mean not getting data to the (older) APM server version if there is a compat issue.

File: index.js [hunk 1]
```diff
@@ -1014,6 +1024,112 @@ function getChoppedStreamHandler (client, onerror) {
...
+ * Fetch the APM Server version and set `this._apmServerVersion`.
+ * https://www.elastic.co/guide/en/apm/server/current/server-info.html
+ *
+ * If fetching/parsing fails then the APM server version will be set to `null`
+ * to indicate "unknown version".
+ */
+Client.prototype._fetchApmServerVersion = function () {
+  const self = this
+
+  const setVerUnknownAndNotify = (errmsg) => {
```
Comment (astorm) [inline index.js:1055]: nit/question: With the `() => {}` style function the use of `self` shouldn't be necessary.  Any particular reason/mental-algorithm for making this an arrow function?

**Review by astorm — APPROVED** [review]
- I see the client now has a private property/member-variable named _apmServerVersion. 
- I see the value for _apmServerVersion will be read from the configuration provided when instantiating the client.
- If no configuration is provided when instantiating the client or the provided values is false-ish we'll call `Client._fetchApmServerVersion()` to fetch the value asyncronously from APM Server via an HTTP request
- A new method on the client, `supportsKeepingUnsampledTransaction`, uses the values to determine whether the APM Server versions supports keeping unsampled transactions.
- I see the functionality pulls in a new module dependency (`semver`,) for the client.  This seems reasonable, especially since we already use this module in the agent.
- The only potential hiccup I see is the period of time between instantiating a client and the request to APM server completing, but we appear to account for that by defaulting to the old, pre-APM Server 8 behavior.  

Approving.

File: index.js (hunk 1 above)
Comment (trentm) [reply index.js:1055]: This old dog is slowing attempting to master new tricks? ;)
I can drop the `self`ing.

**Comment by trentm** [comment]
elastic-apm-http-client@10.4.0 published
v10.4.0 tagged

**Pull request merged**